x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
         ensureOpen();
         return tracker.getAbsentRangeWithin(start, end);
     }
+
+    // used in tests
+    boolean isFSynced() {
+        return fsynced.get();
+    }
+
+    /**
+     * Ensure that all ranges of data written to the cache file are written to the storage device that contains it. This method returns the
+     * list of all successfully written ranges of data since the creation of the cache file.
+     *
+     * @return the list of ranges of data available in cache at the time this method is invoked
+     * @throws IOException                       if the cache file failed to be fsync
+     * @throws AlreadyClosedException            if the cache file is evicted
+     * @throws java.nio.file.NoSuchFileException if the cache file does not exist
+     */
+    public List<Tuple<Long, Long>> fsync() throws IOException {
+        ensureOpen();
+        if (refCounter.tryIncRef()) {
+            try {
+                // Capture the completed ranges before fsyncing; ranges that are completed after this point won't be considered as
+                // persisted on disk by the caller of this method, even if they are fully written to disk at the time the file
+                // fsync is effectively executed
+                final List<Tuple<Long, Long>> completedRanges = tracker.getCompletedRanges();
+                assert completedRanges != null;
+
+                if (fsynced.compareAndSet(false, true)) {
```

Comment:
I think this is OK, but your comment and your suggestion about returning an empty set of ranges if the cache file got evicted made me change a bit the semantic of this new fsync method. 

I pushed d55278ebf5c6f6ccac3e7af06f3124700215a347 which makes the method return a non-empty set of completed ranges only if the cache file was actually fsynced on disk; otherwise it returns an empty collection. This should address the case of concurrent threads and also simplify the logic to know if new ranges of data were effectively synchronized on disk or not.